recipe/gen_patch_json.py

```diff
@@ -808,6 +808,22 @@ def _gen_new_index_per_key(repodata, subdir, index_key):
         if i >= 0:
             deps[i] = "cudatoolkit >=11.2,<12.0a0"
 
+        # cuda-python 11.7.1 is ABI-breaking, currently it impacts rmm & numba
+        if record_name == "rmm":
```

Comment:
Why are we checking only `rmm`? Potentially there could be other affected packages.

Would suggest following a pattern like [this one]( https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/blob/ebab36266e6619116e8322c3e0cf8908ed4162ea/recipe/gen_patch_json.py#L768-L770 ).

Also we may want to [bound this with a timestamp]( https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/blob/ebab36266e6619116e8322c3e0cf8908ed4162ea/recipe/gen_patch_json.py#L774 ) so we limit this to the packages preceding the `run_export`.

Reply:
I determined the impact by searching all conda-forge codebase for `cuda-python`, and concluded that only `rmm` and `numba` are impacted. Wouldn't using precise targets as opposed to a general for loop + if checks reduce the bot's runtime?

Reply:
Not really. The patching occurs as part of the repodata package build in this feedstock. The result is a generated JSON file that we ship. So bots don't need to regenerate this (they just read that JSON).

As part of the build in this feedstock, we loop over all the packages anyways. So this only comes down to how much work it is to check for the package, which is pretty trivial. The timestamp further filters this down. The bigger time sink is if we get this wrong and have to iterate on changing this again.

Reply:
@jakirkham I don't think there's a way to differentiate between packages that only use the cuda-python Python API vs use the Cython API. If only using the Python API then you can float between versions without issue,

Reply:
What issue occurs if we pin both of them?

My concern is this filter feels fragile. Spending time doing this by hand is also cumbersome. It may also entail repeated intervention if we missed things.

If this issue occurs again in another version, we may want to borrow from the current code, but as it is formulated we can't do that.

Reply:
@jakirkham Unfortunately the suggested pattern does not work for `rmm` because only one version (22.08) was built with CUDA Python 11.7.1. As a result, it must be pinned differently from prior versions. 

For `numba`, I think it only uses CUDA Python's Python APIs (@gmarkall to confirm) so we probably don't even need to patch it.

Reply:
Sorry, but am not following.

Why are we adding the `run_export` line to `cuda-python` for all packages it builds with ( https://github.com/conda-forge/cuda-python-feedstock/pull/17 ), if we only want this constraint applied to some packages? It seems if we want this applied to only some packages, those packages should be adding `pin_compatible` instead (like what we do with `numpy`).

Please clarify if I'm missing something (again have been out for a while and am only recently catching up on this discussion)

Reply:
Hi @jakirkham We added `run_export` to the `cuda-python` recipe because CUDA Python 11.7.0 and 11.7.1+ are ABI incompatible. `pin_compatible` in each individual downstream works, but only in each individual downstream feedstock that depends on `cuda-python`, and they need to be aware of this ABI problem first. I think it is too much to ask.

We do NOT only want this constraint applied to some packages. Again, the patch is done here in this particular way because only two packages are impacted by the ABI incompatibility issue.

Reply:
> For `numba`, I think it only uses CUDA Python's Python APIs (@gmarkall to confirm) so we probably don't even need to patch it.

I only just noticed this, apologies - indeed Numba only uses the Python API.

Reply:
Thanks, Graham. Removed the Numba patch in e66febc.